pkg/tbtc/node.go

```diff
@@ -304,12 +336,22 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
 				)
 				defer dkgResultSubscription.Unsubscribe()
 
+				// Set up the publication stop signal that should allow to
+				// perform all the result-signing-related actions and
+				// handle the worst case when the result is submitted by the
+				// last group member.
+				publicationCtx, cancelPublicationCtx := context.WithTimeout(
+					context.Background(),
+					dkgResultSigningDuration+
+						(time.Duration(chainConfig.GroupSize)*dkgResultSubmissionDelayStep),
+				)
+				defer cancelPublicationCtx()
```

Comment:
Don't take it for granted, I am still testing/experimenting but I think this construction of a context does not work well in all cases. I have been running tests locally with 4 clients. 3 clients managed to submit the result successfully. One client got stuck on `resultSigningState` and timed out after 20 minutes.

Looking at the logs, I think everyone completed work before client 4 started it, and given the original message did not get retransmitted, the client got stuck:

From the client 3:
``` 
2022-11-04T14:52:29.002+0100	INFO	keep-tbtc	state/async_machine.go:179	[member:12,state:*dkg.resultSigningState] transitioned to new state	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
2022-11-04T14:52:29.095+0100	INFO	keep-tbtc	state/async_machine.go:179	[member:97,state:*dkg.resultSigningState] transitioned to new state	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
2022-11-04T14:52:29.103+0100	INFO	keep-tbtc	state/async_machine.go:179	[member:12,state:*dkg.signaturesVerificationState] transitioned to new state{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
(...)
2022-11-04T14:52:40.009+0100	INFO	keep-tbtc	state/async_machine.go:148	[member:96,state:*dkg.signaturesVerificationState] transitioning to a new state	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
(...)
2022-11-04T14:52:40.114+0100	INFO	keep-tbtc	state/async_machine.go:148	[member:96,state:*dkg.resultSubmissionState] transitioning to a new state	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
(...)
2022-11-04T14:52:40.331+0100	INFO	keep-tbtc	state/async_machine.go:117	[member:19,state:*dkg.resultSubmissionState] reached final state	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
```

From the client 4:
```
2022-11-04T14:52:29.102+0100	INFO	keep-tbtc	state/async_machine.go:148	[member:78,state:*dkg.resultSigningState] transitioning to a new state	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
2022-11-04T14:52:29.102+0100	INFO	keep-tbtc	state/async_machine.go:179	[member:78,state:*dkg.resultSigningState] transitioned to new state	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
(...)
2022-11-04T15:14:08.959+0100	WARN	keep-tbtc	tbtc/node.go:369	[member:29] DKG result publication process failed [context deadline exceeded]	{"seed": "0x63bcb8b9e18ce096aacd89b73bad7f4c7da5307507c5a329e31948d6c27bed58"}
```

I think it is the scenario that we are discussing [here](https://github.com/keep-network/keep-core/pull/3392#discussion_r1013854529) and we could solve it the same way - not canceling the context until the timeout. Something like this:

```
publicationCtx, cancelPublicationCtx := context.WithCancel(context.Background())
go func() {
  time.Sleep(
    dkgResultSigningDuration +
      time.Duration(chainConfig.GroupSize)*dkgResultSubmissionDelayStep,
    )
  cancelPublicationCtx()
}()
```

Reply:
After testing, this suggestion looks like a good one. I had multiple scenarios where due to spending more time in the result signing phase, some clients were still waiting for messages while other clients completed the execution by submitting the result:

From client 1/4:
```
2022-11-04T18:13:09.032+0100	INFO	keep-tbtc	state/async_machine.go:179	[member:5,state:*dkg.resultSubmissionState] transitioned to new state	{"seed": "0xf0a34daccf046f255f09723ee923de695f7c4236981a56cc5287125041c25d17"}
```

From client 4/4:
```
2022-11-04T18:13:28.144+0100	INFO	keep-tbtc	dkg/states.go:448	[member:43, resultSigningState] CanTransition() has [99] messages received	{"seed": "0xf0a34daccf046f255f09723ee923de695f7c4236981a56cc5287125041c25d17"}
2022-11-04T18:13:28.144+0100	INFO	keep-tbtc	state/async_machine.go:148	[member:34,state:*dkg.signaturesVerificationState] transitioning to a new state	{"seed": "0xf0a34daccf046f255f09723ee923de695f7c4236981a56cc5287125041c25d17"}
```

Reply:
Here it is https://github.com/keep-network/keep-core/pull/3392/commits/0d176312d6450d92544b7dd19ec24979d5b1851a